### PR TITLE
9.0 binary content url chm

### PR DESCRIPTION
**Pull request by Gorash**

### CODE REVIEW

**Comment by Gorash** [comment]
@KangOl can you check for binary_content because they are may be a security fail

**Comment by KangOl** [comment]
Exactly.
You should not read an url. It may not be a relative url, or event a path on disk.
If it's a url, it should be return as a redirect (301), nothing more.

**Comment by Gorash** [comment]
and if use get_resource_from_path and check if the path doesn't contains ../ ?

**Comment by qsm-odoo** [comment]
Is the new version ok? I could make it more compact but I would lose readability...

File: addons/web/controllers/main.py [hunk 1]
```diff
@@ -506,8 +506,20 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
...
+        if url_match:
+            module = url_match.group(1)
+            module_path = get_module_path(module)
+            module_ressource_path = get_resource_path(module, url_match.group(2))
+            if module_path and module_ressource_path:
+                module_path = os.path.normpath(module_path)
+                module_ressource_path = os.path.normpath(module_ressource_path)
+                if os.path.commonprefix([module_path, module_ressource_path]) == module_path:
+                    status = old_status
+                    content = base64.b64encode(open(module_ressource_path, 'r').read())
```
Comment (KangOl) [inline addons/web/controllers/main.py]: Use a `open` as contextmanager (with)

File: addons/web/controllers/main.py [hunk 2]
```diff
@@ -506,8 +506,20 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
...
+        url_match = re.match("^/(\w+)/(.+)$", obj.url)
+        if url_match:
+            module = url_match.group(1)
+            module_path = get_module_path(module)
+            module_ressource_path = get_resource_path(module, url_match.group(2))
+            if module_path and module_ressource_path:
+                module_path = os.path.normpath(module_path)
+                module_ressource_path = os.path.normpath(module_ressource_path)
+                if os.path.commonprefix([module_path, module_ressource_path]) == module_path:
+                    status = old_status
```
Comment (KangOl) [inline addons/web/controllers/main.py]: Useless

Comment (qsm-odoo) [inline addons/web/controllers/main.py]: Which part? status = old_status is required otherwise we will have a 301 with the file data as content.

Comment (KangOl) [inline addons/web/controllers/main.py]: Right.

**Comment by qsm-odoo** [comment]
(If it is ok to merge, I can merge it)

File: addons/web/controllers/main.py [hunk 3]
```diff
@@ -506,8 +506,21 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
 
     # get content after cache control
     if model == 'ir.attachment' and obj.type == 'url' and obj.url:
```
Comment (KangOl) [inline addons/web/controllers/main.py]: Should be done before at line 480 to handle headers correctly (mimetype should be guessed from filename)

File: addons/web/controllers/main.py [hunk 4]
```diff
@@ -469,12 +469,37 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
...
+            content = obj.url
+    else:
+        content = obj[field] or ''
 
     # filename
     if not filename:
         if filename_field in obj:
             filename = obj[filename_field]
+        elif module_resource_path:
+            filename = module_resource_path.split('/').pop()
```
Comment (KangOl) [inline addons/web/controllers/main.py]: should use `os.sep`

Comment (KangOl) [inline addons/web/controllers/main.py]: or better: `os.path.split`

File: openerp/addons/base/ir/ir_attachment.py [hunk 5]
```diff
@@ -426,6 +426,11 @@ def write(self, cr, uid, ids, vals, context=None):
         # remove computed field depending of datas
         for field in ['file_size', 'checksum']:
             vals.pop(field, False)
+        if 'url' in vals and 'datas_fname' not in vals:
+            vals['datas_fname'] = vals['url'].split('/').pop()
```
Comment (KangOl) [inline openerp/addons/base/ir/ir_attachment.py]: Wrong.

File: openerp/addons/base/ir/ir_attachment.py [hunk 6]
```diff
@@ -454,6 +459,8 @@ def create(self, cr, uid, values, context=None):
         # remove computed field depending of datas
         for field in ['file_size', 'checksum']:
             values.pop(field, False)
+        if 'url' in values and 'datas_fname' not in values:
```
Comment (KangOl) [inline openerp/addons/base/ir/ir_attachment.py]: Still wrong.

File: openerp/addons/base/ir/ir_attachment.py [hunk 7]
```diff
@@ -426,6 +426,11 @@ def write(self, cr, uid, ids, vals, context=None):
         # remove computed field depending of datas
         for field in ['file_size', 'checksum']:
             vals.pop(field, False)
+        if 'url' in vals and 'datas_fname' not in vals:
+            vals['datas_fname'] = os.path.basename(vals['url']) or ''
```
Comment (KangOl) [inline openerp/addons/base/ir/ir_attachment.py]: wrong

**Comment by qsm-odoo** [comment]
OK then just forgot about the two commits in base, I don't think the changes in write are good anyway.
Are the first commit (web_editor) and last commit (web) ok? That's only what I need.

**Comment by KangOl** [comment]
`last_update` should also be changed to `str(os.path.getmtime(module_resource_path))` to have a ETag that follow the real file

**Comment by qsm-odoo** [comment]
Like this?

File: addons/web/controllers/main.py [hunk 8]
```diff
@@ -469,12 +469,40 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
...
+
+    status, headers, content = None, [], None
+
+    # attachment by url check
+    module_resource_path = None
+    if model == 'ir.attachment' and obj.type == 'url' and obj.url:
+        url_match = re.match("^/(\w+)/(.+)$", obj.url)
+        if url_match:
+            module = url_match.group(1)
+            module_path = get_module_path(module)
```
Comment (KangOl) [inline addons/web/controllers/main.py]: Should ensure it contains a trailing `/` (use `os.sep`)

File: addons/web/controllers/main.py [hunk 9]
```diff
@@ -469,12 +469,40 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
...
+            module_resource_path = get_resource_path(module, url_match.group(2))
+            if module_path and module_resource_path:
+                module_path = os.path.normpath(module_path)
+                module_resource_path = os.path.normpath(module_resource_path)
+                if os.path.commonprefix([module_path, module_resource_path]) == module_path:
```
Comment (KangOl) [inline addons/web/controllers/main.py]: Uses `startswith` which should be faster.

File: addons/web/controllers/main.py [hunk 10]
```diff
@@ -469,12 +469,40 @@ def binary_content(xmlid=None, model='ir.attachment', id=None, field='datas', un
...
+                    with open(module_resource_path, 'r') as f:
+                        content = base64.b64encode(f.read())
+                        last_update = str(os.path.getmtime(module_resource_path))
```
Comment (KangOl) [inline addons/web/controllers/main.py]: unindent this line

File: addons/web/controllers/main.py (hunk 8 above)
Comment (KangOl) [inline addons/web/controllers/main.py]: You can use `os.path.join(module_path, '')`

Comment (KangOl) [inline addons/web/controllers/main.py]: Leave a comment on why you do this

File: addons/web/controllers/main.py (hunk 9 above)
Comment (qsm-odoo) [inline addons/web/controllers/main.py]: So no commonprefix anymore ?

File: addons/web/controllers/main.py (hunk 8 above)
Comment (qsm-odoo) [inline addons/web/controllers/main.py]: This trailing '/' will be removed by os.path.normpath just below, so useless

**Comment by qsm-odoo** [comment]
Should be ok by now

**Comment by KangOl** [comment]
Rebased on 9.0

**Pull request closed without merging**